v1/dependencies/deps.go

```diff
@@ -168,25 +168,27 @@ func virtual(compiler *ast.Compiler, x interface{}, virtualRefs *dependencies) e
 }
 
 type dependencies struct {
-	refs         *util.HashMap
-	visitedRules *util.HashMap
+	refs         *util.HasherMap[ast.Ref, ast.Ref]
+	visitedRules *util.TypedHashMap[*ast.Rule, *ast.Rule]
 }
 
 func newRefSet() *dependencies {
 	return &dependencies{
-		refs: util.NewHashMap(func(a, b util.T) bool {
-			return a.(ast.Ref).Equal(b.(ast.Ref))
-		}, func(a util.T) int {
-			return a.(ast.Ref).Hash()
-		}),
-		visitedRules: util.NewHashMap(func(a, b util.T) bool {
-			return a.(*ast.Rule).Equal(b.(*ast.Rule))
-		}, func(a util.T) int {
-			return a.(*ast.Rule).Ref().Hash()
-		}),
+		refs: util.NewHasherMap[ast.Ref, ast.Ref](ast.RefEqual),
+		visitedRules: util.NewTypedHashMap[*ast.Rule, *ast.Rule](
+			(*ast.Rule).Equal,
+			nil,
+			ruleHash,
+			nil,
+			nil,
```

Comment:
This is the one place we can't use HasherMap, as Rule doesn't implement Hash. A good reminder however of why the HasherMap is nicer to work with :)